algorithms/nautilus/grammartec/rule.cpp

```diff
@@ -0,0 +1,409 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+/**
+ * @file rule.hpp
+ * @brief Grammar rules and parser/unparser
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ *
+ * @details This file defines Rule class.
+ *          Rule parses terminal and nonterminal symbols written in text.
+ *          It can also unparse a rule into text.
+ */
+#include <iomanip>
+#include <iterator>
+#include <regex>
+#include <sstream>
+#include <string>
+#include "fuzzuf/algorithms/nautilus/grammartec/context.hpp"
+#include "fuzzuf/algorithms/nautilus/grammartec/rule.hpp"
+#include "fuzzuf/algorithms/nautilus/grammartec/tree.hpp"
+#include "fuzzuf/exceptions.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+
+namespace fuzzuf::algorithm::nautilus::grammartec {
+
+/**
+ * @fn
+ * Constructor of Rule class.
+ * This constructor takes a plain-text rule.
+ * @brief Construct Rule from format
+ * @param (ctx) Context
+ * @param (nonterm) String representation of nonterminal symbol
+ * @param (format) Format string (expression) of the rule for @p nonterm
+ *
+ * @details Nonterminal symbol must follow `{ABC:xyz}` format
+ *          where `ABC` indicates the nonterminal and `:xyz` is
+ *          an optional comment.
+ *          The format string is what the nonterminal should be
+ *          replaced with.
+ *          You can use any nonterminal symbols in this format.
+ */
+Rule::Rule(Context& ctx, const std::string& nonterm, const std::string& format) {
+  std::vector<RuleChild> children = Rule::Tokenize(format, ctx);
+  std::vector<NTermID> nonterms;
+
+  // Filter only NTerm items from children
+  for (RuleChild child: children) {
+    if (std::holds_alternative<NTerm>(child.value())) {
+      nonterms.emplace_back(std::get<NTerm>(child.value()));
+    }
+  }
+
+  _rule = PlainRule{ctx.AquireNTID(nonterm), children, nonterms};
+}
+
+/**
+ * @fn
+ * Describe this rule into a human-readable string.
+ * @brief Describe this rule
+ * @param (ctx) Context
+ * @return Human-readable string of this rule
+ */
+std::string Rule::DebugShow(Context& ctx) const {
+  return std::visit([&ctx](auto r) { return r.DebugShow(ctx); }, _rule);
+}
+
+/**
+ * @fn
+ * Remove every backslash before `{` and `}`
+ * @brief Unescape "\{" and "\}" in a string
+ * @param (bytes) String to unescape
+ * @return Unescaped string
+ */
+std::string Rule::Unescape(const std::string& bytes) {
+  if (bytes.size() < 2) {
+    // Nothing to escape
+    return bytes;
+  }
+
+  /* Convert "\{" into "{" and "\}" into "}" */
+  std::string res;
+  size_t i;
+  for (i = 0; i < bytes.size(); i++) {
+    if (bytes[i] == '\\' && bytes[i+1] == '{') {
```

Comment:
This causes a buffer overrun. Please use `bytes.size()-1`.

Reply:
Fixed in [82c195c](https://github.com/fuzzuf/fuzzuf/pull/23/commits/82c195c436c8cfb04c86daa5abc249ced0e6b126)